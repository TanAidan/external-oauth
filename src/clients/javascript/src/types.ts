declare namespace Types {
  export interface ConnectOptions {
    authId?: string
    setupId?: string
    params?: any
  }

  export interface ConnectSuccess {
    authId: string
  }

  export interface ConnectError extends Error {}

  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
