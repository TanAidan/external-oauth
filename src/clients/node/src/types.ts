declare namespace Types {
  export interface IntegrationOptions {
    authId?: string
    setupId?: string
  }

  export type RequestHeaders = Record<string, string | number | undefined>
  export type RequestMethod = 'GET' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH'

  export type RequestQueryString = Record<string, string | number>
  export interface RequestParameters {
    headers?: RequestHeaders
    query?: RequestQueryString
    body?: any
  }
}

export default Types
