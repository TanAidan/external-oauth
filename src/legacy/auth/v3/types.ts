import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
