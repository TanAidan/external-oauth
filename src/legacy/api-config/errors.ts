import { UserError } from '../errors'

export class InvalidApiConfig extends UserError {
  constructor(label: string, subAttribute: string | undefined, variableName: string) {
    const subText = subAttribute ? ` '${subAttribute}'` : ''
    super(
      `The API's ${label}${subText} has been configured with an invalid variable \${${variableName}}.`,
      422,
      'INVALID_API_CONFIG'
    )
  }
}

export class MissingApiConfigHeader extends UserError {
  constructor(label: string, subAttribute: string | undefined, variableName: string) {
    const subText = subAttribute ? ` '${subAttribute}'` : ''
    const header = variableName.split('.')[1]
    super(
      `The API's ${label}${subText} configuration requires the '${header}' header to be passed with each request.`,
      400,
      'MISSING_API_CONFIG_HEADER'
    )
  }
}

export class MissingApiConfigConnectParam extends UserError {
  constructor(label: string, subAttribute: string | undefined, variableName: string) {
    const subText = subAttribute ? ` '${subAttribute}'` : ''
    const param = variableName.split('.')[1]
    super(
      // tslint:disable-next-line:max-line-length
      `The API's ${label}${subText} configuration requires the '${param}' query parameter to be passed when connecting.`,
      400,
      'MISSING_API_CONFIG_HEADER'
    )
  }
}
