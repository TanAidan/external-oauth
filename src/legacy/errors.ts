import { TBackendRequestV4 } from '../types'

export abstract class CustomError extends Error {
  constructor(message: string) {
    super(message)

    Object.setPrototypeOf(this, new.target.prototype)
    this.name = this.constructor.name

    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, this.constructor)
    } else {
      this.stack = new Error(message).stack
    }
  }
}

export abstract class UserError extends CustomError {
  constructor(message: string, public statusCode: number, public code: string) {
    super(message)
  }
}

export class GenericFunctionError extends UserError {
  constructor(req: TBackendRequestV4, e: Error & { requestId?: string }) {
    super(`Error while executing function`, 422, 'FUNCTION_ERROR')
  }
}

function timeOut(error: string, payload: string): boolean {
  const unhandled = error === 'Unhandled'
  const errorMessage = JSON.parse(payload).errorMessage.toString()
  const timedOut = unhandled && /Task timed out/.test(errorMessage)

  return timedOut
}

export class UnhandledFunctionError extends UserError {
  constructor({ FunctionError, Payload }: { FunctionError: string; Payload: string }) {
    if (timeOut(FunctionError, Payload)) {
      super(`Integration execution timed out`, 504, 'INTEGRATION_EXECUTION_TIMED_OUT')
    } else {
      super(`Error while processing integration: ${Payload} `, 422, 'UNHANDLED_FUNCTION')
    }
  }
}
export class MissingParameter extends UserError {
  constructor(name: string) {
    super(`'${name}' must be set`, 400, 'MISSING_PARAMETER')
  }
}

export class InvalidBuid extends UserError {
  constructor(buid: string) {
    super(
      // tslint:disable-next-line:max-line-length
      `No API found with alias/buid '${buid}', please refer to this link for further information: https://docs.bearer.sh/dashboard/apis#how-to-add-a-new-api`,
      422,
      'INVALID_BUID'
    )
  }
}

export class FunctionNotFound extends UserError {
  constructor(functionName: string) {
    super(`Could not find function '${functionName}'. Perhaps url is incorrect?`, 404, 'FUNCTION_NOT_FOUND')
  }
}

export class SetupDetailsNotFound extends UserError {
  constructor({ clientId, buid, setupId }: { clientId: string; buid: string; setupId: string }) {
    super(
      `Setup details not found for clientId '${clientId}', buid '${buid}' and setupId '${setupId}'`,
      422,
      'SETUP_DETAILS_NOT_FOUND'
    )
  }
}

export class RateLimitReached extends UserError {
  constructor() {
    super(
      `Your request couldn't be performed because you have reached Bearer's rate limit.
More information on our rate limiting here: https://docs.bearer.sh/rate-limiting`,
      429,
      'API_RATE_LIMIT_REACHED'
    )
  }
}

export class UnauthorizedAgentConfigAccess extends UserError {
  constructor() {
    super('Unauthorized invalid secret key', 401, 'UNAUTHORIZED_AGENT_CONFIG_ACCESS')
  }
}
