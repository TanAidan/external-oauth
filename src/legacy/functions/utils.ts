// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
