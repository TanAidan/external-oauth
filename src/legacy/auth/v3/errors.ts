import { UserError } from '../../errors'
import { EAuthType } from './types'
import { Request } from 'express'
import { GrantType } from './strategies/oauth2'

export class NoAuthInProgress extends UserError {
  constructor() {
    super(
      `The callback was invoked whilst no authentication was in progress. ` +
        `Either too much time has elapsed between starting authentication and the callback being invoked, ` +
        `or the callback was invoked without first calling the integration's authentication endpoint. Please try again`,
      422,
      'NO_AUTH_IN_PROGRESS'
    )
  }
}
export class InvalidAuthType extends UserError {
  constructor(authType: string) {
    const allowedValues = Object.values(EAuthType)
      .map(authType => `'${authType}'`)
      .join(', ')
    super(
      `Invalid authType '${authType}'. Please check the integration's auth.config.json file. ` +
        `Allowed values are ${allowedValues}`,
      422,
      'INVALID_AUTH_TYPE'
    )
  }
}

export class InvalidGrantType extends UserError {
  constructor(grantType: string) {
    const allowedValues = Object.values(GrantType)
      .map(grantType => `'${grantType}'`)
      .join(', ')
    super(
      `Invalid grantType '${grantType}'. Please check the API's configuration. ` +
        `Allowed values are ${allowedValues}`,
      422,
      'INVALID_GRANT_TYPE'
    )
  }
}

export class OAuthOnlyEndpoint extends UserError {
  constructor(authType: string) {
    super(
      "This endpoint is only for 'OAUTH1' and 'OAUTH2' auth types " +
        `but the integration is configured with authType '${authType}'`,
      422,
      'OAUTH_ONLY_ENDPOINT'
    )
  }
}

export class AuthenticationFailed extends UserError {
  readonly error: string
  readonly errorDescription?: string

  static fromOAuthRequest(req: Request, info: any) {
    if (req.query.error) {
      return new AuthenticationFailed(req.query)
    }

    return new AuthenticationFailed(info)
  }

  constructor(details: any) {
    super(`Authentication failed:\n${JSON.stringify(details, undefined, 2)}`, 403, 'AUTHENTICATION_FAILED')

    if (details.error) {
      this.error = details.error
      this.errorDescription = details.error_description
    } else {
      this.errorDescription = this.error = JSON.stringify(details)
    }
  }
}

export class AccessTokenExpired extends UserError {
  constructor() {
    super(
      `The access token associated with your Auth Id has expired and token refreshing is not supported for \
this API. Please reconnect to obtain a new access token.

Consult the API provider's documentation to determine whether refresh tokens can be enabled for this API`,
      403,
      'ACCESS_TOKEN_EXPIRED'
    )
  }
}

export class InconsistentSetupId extends UserError {
  constructor() {
    super(
      'The provided setupId does not match the setupId used when connecting. ' +
        'Please reconnect using this setupId, or retry this call with the ' +
        'setupId corresponding to the authId',
      422,
      'INCONSISTENT_SETUP_ID'
    )
  }
}

export class InvalidAuthId extends UserError {
  constructor(buid: string, authId: string) {
    super(
      `No auth details were found using authId '${authId}' for API '${buid}'

Please try again with a valid authId or connect with this authId`,
      404,
      'INVALID_AUTH_ID'
    )
  }
}

export class MissingAuthId extends UserError {
  constructor(buid: string) {
    super(
      `You must supply an authId to use the '${buid}' API

Please try again with a valid authId.

See the following link for information on how to obtain an authId: https://docs.bearer.sh/faq/connect-button`,
      401,
      'MISSING_AUTH_ID'
    )
  }
}

export class CredentialsNotConfigured extends UserError {
  constructor(buid: string) {
    super(
      `Missing credentials for the '${buid}' API

Please configure the credentials in the dashboard and try again.

See the following link for information: https://docs.bearer.sh/dashboard/apis`,
      422,
      'CREDENTIALS_NOT_CONFIGURED'
    )
  }
}
