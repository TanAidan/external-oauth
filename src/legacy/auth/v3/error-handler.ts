import { TErrorHandlerRequest } from './types'
import { Response, NextFunction } from 'express'
import { AuthenticationFailed, NoAuthInProgress } from './errors'
import { isEmpty } from '../../functions/utils'

const respondWithOAuthError = (
  res: Response,
  { statusCode, code, message }: { statusCode: number; code: string; message: string }
) => {
  res.header('Content-Type', 'text/html')
  res.status(statusCode)
  res.render('auth/error', { error: `${code}: ${message}` })
}

const respondWithCallbackError = (req: TErrorHandlerRequest, res: Response, err: AuthenticationFailed) => {
  res.header('Content-Type', 'text/html')
  res.status(err.statusCode)
  res.render('auth/callback', {
    authId: req.authId,
    error: err.error,
    error_description: err.errorDescription,
    integrationUuid: req.buid
  })
}

const respondWithCallbackPlaceholder = (
  req: TErrorHandlerRequest,
  res: Response,
  { statusCode, code, message }: { statusCode: number; code: string; message: string }
) => {
  res.header('Content-Type', 'text/html')
  res.status(statusCode)
  res.render('auth/placeholder', {
    error: `${code}: ${message}`,
    callbackUrl: req.originalUrl,
    method: req.method
  })
}

export const errorHandler = (err: any, req: TErrorHandlerRequest, res: Response, next: NextFunction) => {
  if (res.headersSent) {
    return next(err)
  }

  if (err.statusCode) {
    if (err instanceof AuthenticationFailed) {
      return respondWithCallbackError(req, res, err)
    }

    if (err instanceof NoAuthInProgress && isEmpty(req.query)) {
      return respondWithCallbackPlaceholder(req, res, err)
    }

    return respondWithOAuthError(res, err)
  }
  respondWithOAuthError(res, {
    statusCode: 500,
    code: 'INTERNAL_ERROR',
    message: 'Encountered an unexpected error. Please contact support'
  })
}
