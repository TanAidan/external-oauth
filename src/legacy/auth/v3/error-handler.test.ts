import { ErrorMiddlewareTestHarness } from '../../../../tests/utils'
import { TErrorHandlerRequest } from './types'
// @ts-ignore
import { errorHandler } from './error-handler'
import { UserError } from '../../../errors'
import { AuthenticationFailed, NoAuthInProgress } from './errors'

class TestUserError extends UserError {
  constructor() {
    super('test error', 422, 'TEST_ERROR')
  }
}

class TestNoAuthInProgressError extends NoAuthInProgress {}

describe('errorHandler', () => {
  const log = jest.fn()
  const setLogLevel = jest.fn()

  const authenticationFailedError = new AuthenticationFailed({ error: 'oops_code', error_description: 'some error' })
  const nonUserError = new Error('oh noes')

  const setup = ({ withLogger = true, sendResponse = false, url = '' } = {}) =>
    new ErrorMiddlewareTestHarness<TErrorHandlerRequest>({
      configureRequest: (req, res) => {
        req.buid = 'test-buid'
        req.authId = 'test-auth-id'

        if (url.length > 0) {
          req.originalUrl = url
        }

        if (sendResponse) {
          res.send()
        }
      },
      testMiddleware: errorHandler
    })

  beforeEach(() => {
    log.mockClear()
    setLogLevel.mockClear()
  })

  describe('when NoAuthInProgress error is returned to request WITHOUT query params', () => {
    const test = setup({ url: 'v2/auth/callback' })

    it('renders the `callback-url-request-error` template with the expected parameters', async () => {
      await test
        .get(new TestNoAuthInProgressError())
        .expect('Content-Type', 'text/html; charset=utf-8')
        .expect(422)
        .type('text/html')

      expect(test.view).toMatchSnapshot()
    })

    it("doesn't pass on the error", async () => {
      await test.get(new TestUserError()).expect(422)

      expect(test.err).toBeUndefined()
    })
  })

  describe('when NoAuthInProgress error is returned to request WITH query params', () => {
    const test = setup({ url: 'v2/auth/callback' })

    it('renders the `oauth-error` template with the expected parameters', async () => {
      await test
        .get(new TestNoAuthInProgressError())
        .query('?hello=world')
        .expect('Content-Type', 'text/html; charset=utf-8')
        .expect(422)
        .type('text/html')

      expect(test.view).toMatchSnapshot()
    })

    it("doesn't pass on the error", async () => {
      await test
        .get(new TestUserError())
        .query('?hello=world')
        .expect(422)

      expect(test.err).toBeUndefined()
    })
  })

  describe('when the error is a UserError', () => {
    describe('when there is a logger on the request', () => {
      it("logs the error to the request's logger with ERROR level", async () => {
        const test = setup()

        await test.get(new TestUserError()).expect(422)

        expect(log).toMatchSnapshot()
      })
    })

    describe('when the error is an AuthenticationFailed error', () => {
      it('renders the `callback` template with the expected parameters', async () => {
        const test = setup()

        await test
          .get(authenticationFailedError)
          .expect('Content-Type', 'text/html; charset=utf-8')
          .expect(403)
          .type('text/html')

        expect(test.view).toMatchSnapshot()
      })

      it("doesn't pass on the error", async () => {
        const test = setup()

        await test.get(authenticationFailedError).expect(403)

        expect(test.err).toBeUndefined()
      })
    })

    describe('when the error is NOT an AuthenticationFailed error', () => {
      it('renders the `oauth-error` template with the expected parameters', async () => {
        const test = setup()

        await test
          .get(new TestUserError())
          .expect('Content-Type', 'text/html; charset=utf-8')
          .expect(422)
          .type('text/html')

        expect(test.view).toMatchSnapshot()
      })

      it("doesn't pass on the error", async () => {
        const test = setup()

        await test.get(new TestUserError()).expect(422)

        expect(test.err).toBeUndefined()
      })
    })
  })

  describe('when the error is NOT a UserError', () => {
    it('renders the `oauth-error` template with the expected parameters', async () => {
      const test = setup()

      await test
        .get(nonUserError)
        .expect('Content-Type', 'text/html; charset=utf-8')
        .expect(500)
        .type('text/html')

      expect(test.view).toMatchSnapshot()
    })

    it('passes on the error', async () => {
      const test = setup()

      await test.get(nonUserError).expect(500)

      expect(test.err).toBe(nonUserError)
    })
  })

  describe('when headers have already been sent', () => {
    it("doesn't render anything", async () => {
      const test = setup({ sendResponse: true })

      await test.get(nonUserError).expect(200)

      expect(test.view).toBeUndefined()
    })

    it('passes on the error', async () => {
      const test = setup({ sendResponse: true })

      await test.get(nonUserError).expect(200)

      expect(test.err).toBe(nonUserError)
    })
  })
})
