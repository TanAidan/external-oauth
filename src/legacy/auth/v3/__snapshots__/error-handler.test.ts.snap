// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorHandler when NoAuthInProgress error is returned to request WITH query params renders the \`oauth-error\` template with the expected parameters 1`] = `
Object {
  "name": "oauth-error",
  "options": Object {
    "error": "NO_AUTH_IN_PROGRESS: The callback was invoked whilst no authentication was in progress. Either too much time has elapsed between starting authentication and the callback being invoked, or the callback was invoked without first calling the integration's authentication endpoint. Please try again",
  },
}
`;

exports[`errorHandler when NoAuthInProgress error is returned to request WITHOUT query params renders the \`callback-url-request-error\` template with the expected parameters 1`] = `
Object {
  "name": "callback-url-request-error",
  "options": Object {
    "callbackUrl": "v2/auth/callback",
    "error": "NO_AUTH_IN_PROGRESS: The callback was invoked whilst no authentication was in progress. Either too much time has elapsed between starting authentication and the callback being invoked, or the callback was invoked without first calling the integration's authentication endpoint. Please try again",
    "method": "GET",
  },
}
`;

exports[`errorHandler when the error is NOT a UserError renders the \`oauth-error\` template with the expected parameters 1`] = `
Object {
  "name": "oauth-error",
  "options": Object {
    "error": "INTERNAL_ERROR: Encountered an unexpected error. Please contact support",
  },
}
`;

exports[`errorHandler when the error is a UserError when the error is NOT an AuthenticationFailed error renders the \`oauth-error\` template with the expected parameters 1`] = `
Object {
  "name": "oauth-error",
  "options": Object {
    "error": "TEST_ERROR: test error",
  },
}
`;

exports[`errorHandler when the error is a UserError when the error is an AuthenticationFailed error renders the \`callback\` template with the expected parameters 1`] = `
Object {
  "name": "callback",
  "options": Object {
    "authId": "test-auth-id",
    "error": "oops_code",
    "error_description": "some error",
    "integrationUuid": "test-alias",
  },
}
`;

exports[`errorHandler when the error is a UserError when there is NOT a logger on the request logs the error to the debug logger 1`] = `[MockFunction]`;

exports[`errorHandler when the error is a UserError when there is a logger on the request logs the error to the request's logger with ERROR level 1`] = `[MockFunction]`;
